### PR TITLE
3722

### PR DESCRIPTION
**Pull request by RichardSmedley**
Threshold Logging from the Java SDK

### CODE REVIEW

File: content/sdk/java/tracing-from-the-sdk.dita [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+<!DOCTYPE topic PUBLIC "-//OASIS//DTD DITA Topic//EN" "topic.dtd">
+<topic id="java-tracing-howto">
+  <title>Tracing from the Java SDK with Couchbase Server</title>
+  <titlealts><navtitle>Tracing from the SDK</navtitle></titlealts>
+  <shortdesc>Use tracing from the SDK to discover timeouts and bottlenecks across the netowrk and cluster.</shortdesc>
+  <body>
+    <p>For our example, we will customize the threshold logging tracer settings with <apiname>CouchbaseEnvironment.Builder</apiname>.  By default it 
+      will log every minute (if something is found) and only sample the top 10 slowest operations. The default threshold for key/value operation is 500 milliseconds. We shall
+      set them so low that almost everything gets logged - not something you should do in production!<codeblock outputclass="language-java">Tracer tracer = ThresholdLogTracer.create(ThresholdLogReporter.builder()
+  .kvThreshold(1) // 1 micros
```
Comment (daschl) [inline content/sdk/java/tracing-from-the-sdk.dita]: this is now also 1, TimeUnit.MICROSECONDS similar to the inerval

File: content/sdk/java/tracing-from-the-sdk.dita [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  .sampleSize(Integer.MAX_VALUE)
+  .pretty(true) // pretty print the json output in the logs
+  .build());</codeblock>This will set our threshold for key/value operations to one microsecond, log the found operations every second, and set the sample size to a very large value so everything will be logged.</p>
+
+    <p>With these configs in place we are ready to run some operations. Below, we read some documentss from the travel-sample bucket and, if found, write them back with an <codeph>upsert</codeph> - giving us both read and write operations to log.<codeblock outputclass="language-java">// Connect
+CouchbaseEnvironment env = DefaultCouchbaseEnvironment.builder()
+    .tracer(tracer)
+    .build();
+Cluster cluster = CouchbaseCluster.create(env, "127.0.0.1");
+
```
Comment (daschl) [inline content/sdk/java/tracing-from-the-sdk.dita]: if you do it like this you need to run cluster.authenticate("username", "password"); too

File: content/sdk/java/tracing-from-the-sdk.dita [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    bucket.upsert(doc);
+  }
+}
+ 
+Thread.sleep(TimeUnit.MINUTES.toMillis(1));</codeblock></p>
+
+    <p>Run the code, and you will see something like the following in the logs:<codeblock outputclass="language-json">Apr 04, 2018 9:42:57 AM com.couchbase.client.core.tracing.ThresholdLogReporter logOverThreshold
+WARNING: Operations over threshold: [ {
+  "top" : [ {
+    "server_us" : 8,
```
Comment (daschl) [inline content/sdk/java/tracing-from-the-sdk.dita]: field names of this JSON have changed, a simple copy replace should be okay (see below for the new names) that will ship with GA

File: content/sdk/java/tracing-from-the-sdk.dita [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	"decode_us" : 1203,
+	"dispatch_us" : 947,
+	"remote_address" : "127.0.0.1:11210",
+	"total_us" : 1525
+}</codeblock></p>
+    <p>This tells us the following:<ul>
+      <li><b>total_us:</b> The total time it took to perform the full operation: here around 1.5 milliseconds</li>
+      <li><b>server_us:</b> The server reported that its work performed took 23 microseconds (this does not include network time or time in the buffer before picked up at the cluster)</li>
+      <li><b>decode_us:</b> Decoding the response took the client 1.2 milliseconds</li>
+      <li><b>dispatch_us:</b> The time when the client sent the request and got the response took around 1 millisecond.</li>
```
Comment (daschl) [inline content/sdk/java/tracing-from-the-sdk.dita]: last_dispatch_us

File: content/sdk/java/tracing-from-the-sdk.dita [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      <li><b>local_address:</b> The local socket used for this operation.</li>
```
Comment (daschl) [inline content/sdk/java/tracing-from-the-sdk.dita]: last_local_address

File: content/sdk/java/tracing-from-the-sdk.dita [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+      <li><b>remote_address:</b> The remote socket on the server used for this operation. Useful to figure out which node is affected.</li>
```
Comment (daschl) [inline content/sdk/java/tracing-from-the-sdk.dita]: last_remote_address

File: content/sdk/java/tracing-from-the-sdk.dita [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      <li><b>operation_id:</b> A combination of type of operation and id (in this case the opaque value), useful for diagnosing and troubleshooting in combination with the local_id.</li>
```
Comment (daschl) [inline content/sdk/java/tracing-from-the-sdk.dita]: last_operation_id

File: content/sdk/java/tracing-from-the-sdk.dita [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      <li><b>local_id:</b> With Server 5.5 and later, this id is negotiated with the server and can be used to correlate logging information on both sides in a simpler fashion.</li></ul></p>
```
Comment (daschl) [inline content/sdk/java/tracing-from-the-sdk.dita]: last_local_id

File: content/sdk/java/tracing-from-the-sdk.dita [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+Caused by: java.util.concurrent.TimeoutException
+	... 3 more</codeblock></p>
+
+    <p>The new client does much more, transparently adding even more information:.<codeblock outputclass="language-java">Exception in thread "main" java.lang.RuntimeException: java.util.concurrent.TimeoutException: localId: 2C12AAA6637FB4FF/00000000147B092F, opId: 0x1, local: 127.0.0.1:60389, remote: 127.0.0.1:11210, timeout: 1000us
+	at rx.exceptions.Exceptions.propagate(Exceptions.java:57)
+	at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:463)
+	at rx.observables.BlockingObservable.singleOrDefault(BlockingObservable.java:372)
+	at com.couchbase.client.java.CouchbaseBucket.get(CouchbaseBucket.java:131)
+	at SimpleReadWrite.main(SimpleReadWrite.java:64)
+Caused by: java.util.concurrent.TimeoutException: localId: 2C12AAA6637FB4FF/00000000147B092F, opId: 0x1, local: 127.0.0.1:60389, remote: 127.0.0.1:11210, timeout: 1000us
```
Comment (daschl) [inline content/sdk/java/tracing-from-the-sdk.dita]: this format now also looks different:

i.e.

Exception in thread "main" java.lang.RuntimeException: java.util.concurrent.TimeoutException: {"b":"travel-sample","r":"127.0.0.1:11210","s":"kv","c":"0053FF30C2825267/FFFFFFFFD34A9178","t":10000,"i":"0x1","l":"127.0.0.1:64351"}

File: content/sdk/go/encrypting-using-sdk.dita [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+
+        <section id="decryption">
+          <title>Decryption</title>
+    <p>To decrypt use the <codeph>gocbfieldcrypt.DecryptJsonStruct([]byte, reflect.Type, KeyProvider) ([]byte, error)</codeph> function. For example to get a document and parse it into a struct:<codeblock outputclass="language-go">    var p1Bytes []byte
+    bucket.Get("p1", &amp;p1Bytes)
+ 
+    var p1 Person
+    decBytes, _ := gocbfieldcrypt.DecryptJsonStruct(p1Bytes, reflect.TypeOf(p1), keyStore)
+    json.Unmarshal(decBytes, &amp;p2)</codeblock></p>
```
Comment (chvck) [inline content/sdk/go/encrypting-using-sdk.dita]: Sorry, this needs to be p1

File: content/sdk/go/encrypting-using-sdk.dita [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+       "mypublickey": []byte("!mysecretkey#9^5usdk39d&amp;dlf)03sL"),
+       "myhmackey":   []byte("myauthpassword"),
+    },
+}
+ 
+cluster, _ := gocb.Connect("…")
+cluster.Authenticate(…)
+bucket, _ := cluster.openBucket("…", "")</codeblock></p>
+
+    <p>To apply encryption to the struct use the `gocbfieldcrypt.EncryptJsonStruct([]byte, reflect.Type, KeyProvider) ([]byte, error)` function. For example to create a new person, apply encryption and then upsert the document:<codeblock>    person := Person{
```
Comment (chvck) [inline content/sdk/go/encrypting-using-sdk.dita]: I think that the bit in `` needs wrapping in the correct tags.

File: content/sdk/go/encrypting-using-sdk.dita [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  PUBLIC "-//OASIS//DTD DITA Concept//EN" "concept.dtd">
+<concept id="java-fle">
+  <title>Field Level Encryption from the Go SDK</title>
+  <shortdesc>Field Level Encryption is available in Couchbase Data Platform 5.5, from .NET SDK version 2.6.0.</shortdesc>
+  <conbody>
+        <section id="package">
+          <title>Packaging</title>
+            <p>The Couchbase Go SDK uses the <i>gocbencryption</i> library to provide functionality for encryption and decryption of json []byte data.<codeblock outputclass="language-go">go get github.com/couchbase/gocbencryption</codeblock></p>
+
+    <p>The Couchbase Go Field Level Encryption (FLE) library uses struct tags to specify which field(s) to apply encryption to and which algorithm to use. The struct tag key is <i>cbcrypt</i> and the value is of the form <codeph>"algorithm,publickey,signingkey"</codeph>. Here’s an example struct definition:<codeblock outputclass="language-go">type PersonAddress struct {
```
Comment (chvck) [inline content/sdk/go/encrypting-using-sdk.dita]: Should the bit <i> tag be in a codeph?

File: content/sdk/java/tracing-from-the-sdk.dita (hunk 9 above)
Comment (RichardSmedley) [reply content/sdk/java/tracing-from-the-sdk.dita]: Thanks - changes made.

File: content/sdk/go/encrypting-using-sdk.dita (hunk 12 above)
Comment (RichardSmedley) [reply content/sdk/go/encrypting-using-sdk.dita]: Yes!

**Pull request merged**